libzecale/circuits/fields/fp6_3over2_gadgets.tcc

```diff
@@ -0,0 +1,215 @@
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _c0(v._c0)
+    , _c1(v._c1)
+    , _c2(v._c2)
+{
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T>::Fp6_3over2_variable(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp6T &v,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _c0(pb, v.c0, FMT(annotation_prefix, " c0"))
+    , _c1(pb, v.c1, FMT(annotation_prefix, " c1"))
+    , _c2(pb, v.c2, FMT(annotation_prefix, " c2"))
+{
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T>::Fp6_3over2_variable(
+    libsnark::protoboard<FieldT> &pb,
+    const libsnark::Fp2_variable<Fp2T> &c0,
+    const libsnark::Fp2_variable<Fp2T> &c1,
+    const libsnark::Fp2_variable<Fp2T> &c2,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix), _c0(c0), _c1(c1), _c2(c2)
+{
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T> Fp6_3over2_variable<Fp6T>::operator*(const FieldT &a)
+{
+    return Fp6_3over2_variable<Fp6T>(
+        this->pb,
+        _c0 * a,
+        _c1 * a,
+        _c2 * a,
+        FMT(this->annotation_prefix, " * Fr"));
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T> Fp6_3over2_variable<Fp6T>::operator+(
+    const Fp6_3over2_variable<Fp6T> &a)
+{
+    return Fp6_3over2_variable<Fp6T>(
+        this->pb,
+        _c0 + a._c0,
+        _c1 + a._c1,
+        _c2 + a._c2,
+        FMT(this->annotation_prefix, " + a"));
+}
+
+template<typename Fp6T> void Fp6_3over2_variable<Fp6T>::evaluate() const
+{
+    _c0.evaluate();
+    _c1.evaluate();
+    _c2.evaluate();
+}
+
+template<typename Fp6T>
+void Fp6_3over2_variable<Fp6T>::generate_r1cs_witness(const Fp6T &v)
+{
+    _c0.generate_r1cs_witness(v.c0);
+    _c1.generate_r1cs_witness(v.c1);
+    _c2.generate_r1cs_witness(v.c2);
+}
+
+template<typename Fp6T> Fp6T Fp6_3over2_variable<Fp6T>::get_element() const
+{
+    return Fp6T(_c0.get_element(), _c1.get_element(), _c2.get_element());
+}
+
+// Fp6_3over2_mul_gadget methods
+
+template<typename Fp6T>
+Fp6_3over2_mul_gadget<Fp6T>::Fp6_3over2_mul_gadget(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp6_3over2_variable<Fp6T> &A,
+    const Fp6_3over2_variable<Fp6T> &B,
+    const Fp6_3over2_variable<Fp6T> &result,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _A(A)
+    , _B(B)
+    , _result(result)
+    , _a1b1(
+          pb,
+          A._c1,
+          B._c1,
+          libsnark::Fp2_variable<Fp2T>(pb, FMT(annotation_prefix, " a1b1")),
+          FMT(annotation_prefix, " a1*b1"))
+    , _a2b2(
+          pb,
+          A._c2,
+          B._c2,
+          libsnark::Fp2_variable<Fp2T>(pb, FMT(annotation_prefix, " a2b2")),
+          FMT(annotation_prefix, " a2*b2"))
+    , _a1a2_times_b1b2(
+          pb,
+          A._c1 + A._c2,
+          B._c1 + B._c2,
+          libsnark::Fp2_variable<Fp2T>(
+              pb, FMT(annotation_prefix, " a1a2*b1b2")),
```

Comment:
I would use the variable name as annotation. This makes it easier to track down specific gates in the (JSON) exported R1CS by querying the variable name directly.

Reply:
I've tried to consistently use the variable name for annotations where applicable.  In cases such as this there is no variable name (it's an intermediate fp2_variable), so I've tried to give a description of the meaning.